plugins: add Japanese Support plugin

cyphar opened the pull request:
This PR implements a generic language-support module which can be used by language-specific plugins, and the first such plugin for Japanese text (designed to mimic Yomichan's word lookup system). In order to make this work, several other cleanups were necessary in the code which surrounds the language-support hooks (namely the highlighting system and dictionary lookups). The Japanese support plugin does add more work to the highlight and search systems, in my testing it doesn't seem to be too expensive (unfortunately I can't test this on Android because it requires a rebuilt sdcv binary).

Things left to do:

- [x] (Language Support) Fix the weird issue where after deselecting text, you see a single-word highlight. This only happens for the quick-lookup case (and if you tap "highlight" you end up only highlighting a single word).
- [x] (Language Support) Handle highlight expansions that cross page boundaries. At the moment we simply bail in that case, but since Japanese words can very easily span multiple pages this is a case we need to handle somehow. I need to figure out how to emulate the behaviour hold-and-pan has.
- [x] (Japanese) Look into using `utf8proc` to apply `NFC` normalisation and/or whether we should use the `utf8proc` iterators.
- [x] (Japanese) Stop trying to expand the selection once we hit punctuation.
- [x] (Language Support, Japanese) Improve how this new system matches up with the existing plugin and module system (at the moment it feels like it's not quite fitting in with the rest of the plugin and module system -- for instance we don't register LanguageSupport as a module in the UI for instance).
- [ ] Unit tests.
  - [ ] Japanese
  - [x] Language Support
- [x] (Japanese) Quit early if the selected text is non-CJK text. While some Japanese words use entirely Latin text (NG, CM, and so on) those cases should be handled by the default highlighting system and such words do not need any deinflection.
- [x] Figure out how the menu system for Language Support (and each individual language plugin) should work. Ideally each plugin should have full control over its own submenu from the "Language Support" menu but I'm not quite sure how to do that.
- [x] Test how bad the performance is outside of the emulator.

koreader-base PRs:
- [x] Get koreader/koreader-base#1418 merged which is necessary for `getTextFromXPointers` to do the selection we need.
- [x] Get koreader/koreader-base#1419 merged which is necessary for the `rawSdcv` changes to work properly with what the Japanese segmenter needs.
- [x] Get koreader/koreader-base#1420 merged to be able to normalise strings before Japanese deinflection.

Possible extra features to include:
- [ ] Rather than just passing the candidate words from the plugin, also provide some "tags" that should be shown in the dictionary viewer (in the case of Japanese this would be the list of deinflection reasons which would match Yomichan's style). But for most readers this is not necessary.
- [ ] (Japanese) During deinflection, return the deinflection of any shorter words. This would help make it easier to look up words even if you fat-finger the search, and also would fix the problem that JMDict likes to create entries for words which don't exist in monolingual dictionaries.
- [x] (Japanese) Try to do more of Yomichan's conversions before deinflection, such as collapsing emphatic markers and half-width conversion.
- [x] (Japanese) Allow some kinds of configuration from the main menu:
  - [x] The text scan length limit (the look-ahead limit for expansion).
  - [x] Which set of text conversions to apply when attempting deinflection.

Fixes #4091
/cc @rtega

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/8312)
<!-- Reviewable:end -->

cyphar commented:
> In the end (unless you already did that), may be make the Language support> menu item not plugged when there are no language support plugins that have registered to it, for people (most) that will disable them all.

I still need to figure out how to make the menu work nicely. At the moment it doesn't do anything but it would be nice to be able to have language plugin settings placed there if relevant (though at the moment there are no plugin settings to worry about). However the main issue I had when trying to write the menu system is that it's not clear to me how to create a dynamic menu system where each plugin can have an `addToMainMenu` method that results in that plugins menu being put underneath the `Language Support>` menu. If you have any pointers I'd appreciate it, if not I'll try to figure something out...

poire-z commented:
See these plugins, which can plug their menu item into not-a-top-level menu:
```
plugins/autostandby.koplugin/main.lua:        sorting_hint = "device",
plugins/autosuspend.koplugin/main.lua:        sorting_hint = "device",
plugins/autosuspend.koplugin/main.lua:        sorting_hint = "device",
```
You could have ReaderHighlight create the first level languagesupport= "Language support>", and japanese.koplugin just add itself in ther with `sorting_hint = "languagesupport",`.
We'll have to see if we can make this first level disappear when it ends up being empty.

cyphar commented:
Ah, thanks. I had a feeling `sorting_hint` had something to do with it but I couldn't quite figure out how to use it. I'll take a look at those plugins, thanks!

cyphar commented:
> > However, maybe if we passed a set of callbacks that can be used rather than calling document:... directly would improve things without sacrificing the flexibility currently offered
> 
> Yes, that sounds neat :)

This is now done, it was pretty painless to do actually. Now the plugin gets a table of closures that have `document` as an upvalue and only take XPointers (but if we ever added PDF support, they could take the PDF equivalent).

NiLuJe commented:
As for pure-Lua stuff, please bench vs. FFI, and see what's available in PUC Lua, since it supports UTF-8 natively since something like 5.4 or 5.3, and I *think* most of it is implemented in Lua and not in C?

cyphar commented:
I think that the WTF-8 support in `splitToChars` could be implemented on top of a standard iterator over codepoints because all of its special handling is unscrambling surrogates embedded in UTF-8. As long as the UTF-8 parser doesn't throw errors when a surrogate appears (which it shouldn't since they are valid UTF-8 codepoints) it should work.

But as I said above, for this PR I'll only add the handful of things I need. I'd be happy to do benchmarking and look at a rework of the utf8 handling in koreader in a separate PR. :smile:

poire-z commented:
> If I set self.use_xtext = false in frontend/ui/widget/textboxwidget.lua, the contents of the text box don't update when you try to switch dictionary definitions

Reproduced. Well, it's possible we didn't test a thing or two with use_xtext=false in #7166 and #7212 (pinging @NiLuJe in case you care).

> As far as I can tell, string breaking works as it should (or rather, I couldn't find any instances where the breaking rules were violated when displaying Japanese dictionary definitions) 

But that's good enough for me :)

cyphar commented:
commit 19504f152a198d61003b59528214c5cceb2074f8 ("readerhighlight: remove selected_word and use selected_text everwhere")

Frenzie commented:
@cyphar 
> For one thing, the readerhighlight_spec.lua tests fail. But you can reproduce it in the emulator if you open test/paper.pdf (or any other PDF -- it seems to happen in every PDF I've tried) and enable reflow, then try to hold-and-pan on any word you'll get this crash:

That's exactly what I tried earlier. No crash.

cyphar commented:
I've managed to reproduce the reflowing crashes on `master` so this seems to actually be an existing bug. The errors come from `k2pdfopt_get_native_word_boxes` specifically, but the function is called with the same parameters with and without this PR applied (and now that it always seems to crash on master I can't really figure out what's wrong with it). In practice though, it seems the crashes are a bit of a freak occurrence caused if you enable reflowing with a ridiculously large screen size -- I expect no user is actually going to run into this issue (and the ultimate fix might need to be updating leptonica which requires updating everything else in the mupdf stack).

I'm still looking into the unit test failures though, because those are being triggered without calling `getWordBoxes` (and thus `k2pdfopt_get_native_word_boxes`) at all.

cyphar commented:
So it seems that the issue is that `KoptInterface:getReflowedTextBoxes{FromScratch}` returns nil because there is no cached `kctx_hash` and it doesn't try to recreate one. Here's the diff:

```patch
diff --git a/frontend/document/koptinterface.lua b/frontend/document/koptinterface.lua
index c178d05ebaa0..84e458c8d150 100644
--- a/frontend/document/koptinterface.lua
+++ b/frontend/document/koptinterface.lua
@@ -538,19 +538,22 @@ function KoptInterface:getReflowedTextBoxes(doc, pageno)
     local hash = table.concat(hash_list, "|")
     local cached = DocCache:check(hash)
     if not cached then
+        local kc
         local kctx_hash = hash:gsub("^rfpgboxes|", "kctx|")
         cached = DocCache:check(kctx_hash)
-        if cached then
-            local kc = self:waitForContext(cached.kctx)
-            --kc:setDebug()
-            local fullwidth, fullheight = kc:getPageDim()
-            local boxes, nr_word = kc:getReflowedWordBoxes("dst", 0, 0, fullwidth, fullheight)
-            if not boxes then
-                return nil
-            end
-            DocCache:insert(hash, CacheItem:new{ rfpgboxes = boxes, size = 192 * nr_word }) -- estimation
-            return boxes
+        if not cached then
+            kc = self:getCachedContext(doc, pageno)
+        else
+            kc = self:waitForContext(cached.kctx)
+        end
+        --kc:setDebug()
+        local fullwidth, fullheight = kc:getPageDim()
+        local boxes, nr_word = kc:getReflowedWordBoxes("dst", 0, 0, fullwidth, fullheight)
+        if not boxes then
+            return nil
         end
+        DocCache:insert(hash, CacheItem:new{ rfpgboxes = boxes, size = 192 * nr_word }) -- estimation
+        return boxes
     else
         return cached.rfpgboxes
     end
@@ -566,19 +569,27 @@ function KoptInterface:getNativeTextBoxes(doc, pageno)
     local hash = table.concat(hash_list, "|")
     local cached = DocCache:check(hash)
     if not cached then
+               local kc
         local kctx_hash = hash:gsub("^nativepgboxes|", "kctx|")
         cached = DocCache:check(kctx_hash)
-        if cached then
-            local kc = self:waitForContext(cached.kctx)
-            --kc:setDebug()
-            local fullwidth, fullheight = kc:getPageDim()
-            local boxes, nr_word = kc:getNativeWordBoxes("dst", 0, 0, fullwidth, fullheight)
-            if not boxes then
-                return nil
-            end
-            DocCache:insert(hash, CacheItem:new{ nativepgboxes = boxes, size = 192 * nr_word }) -- estimation
-            return boxes
-        end
+               if not cached then
+                       kc = self:createContext(doc, pageno)
+                       DocCache:insert(kctx_hash, ContextCacheItem:new{
+                               persistent = true,
+                               size = self.last_context_size or self.default_context_size,
+                               kctx = kc,
+                       })
+               else
+                       kc = self:waitForContext(cached.kctx)
+               end
+               --kc:setDebug()
+               local fullwidth, fullheight = kc:getPageDim()
+               local boxes, nr_word = kc:getNativeWordBoxes("dst", 0, 0, fullwidth, fullheight)
+               if not boxes then
+                       return nil
+               end
+               DocCache:insert(hash, CacheItem:new{ nativepgboxes = boxes, size = 192 * nr_word }) -- estimation
+               return boxes
     else
         return cached.nativepgboxes
     end
@@ -596,21 +607,24 @@ function KoptInterface:getReflowedTextBoxesFromScratch(doc, pageno)
     local hash = table.concat(hash_list, "|")
     local cached = DocCache:check(hash)
     if not cached then
+        local reflowed_kc
         local kctx_hash = hash:gsub("^scratchrfpgboxes|", "kctx|")
         cached = DocCache:check(kctx_hash)
-        if cached then
-            local reflowed_kc = self:waitForContext(cached.kctx)
-            local fullwidth, fullheight = reflowed_kc:getPageDim()
-            local kc = self:createContext(doc, pageno)
-            kc:copyDestBMP(reflowed_kc)
-            local boxes, nr_word = kc:getNativeWordBoxes("dst", 0, 0, fullwidth, fullheight)
-            kc:free()
-            if not boxes then
-                return nil
-            end
-            DocCache:insert(hash, CacheItem:new{ scratchrfpgboxes = boxes, size = 192 * nr_word }) -- estimation
-            return boxes
+        if not cached then
+            reflowed_kc = self:getCachedContext(doc, pageno)
+        else
+            reflowed_kc = self:waitForContext(cached.kctx)
         end
+        local fullwidth, fullheight = reflowed_kc:getPageDim()
+        local kc = self:createContext(doc, pageno)
+        kc:copyDestBMP(reflowed_kc)
+        local boxes, nr_word = kc:getNativeWordBoxes("dst", 0, 0, fullwidth, fullheight)
+        kc:free()
+        if not boxes then
+            return nil
+        end
+        DocCache:insert(hash, CacheItem:new{ scratchrfpgboxes = boxes, size = 192 * nr_word }) -- estimation
+        return boxes
     else
         return cached.scratchrfpgboxes
     end
```

This still doesn't fix the test failures, but now the failures are in the assertions (the check that the pop-up is correct is trying to index a nil). I just wanted to post this in case this is obviously wrong and someone wants to scream at me before I get the two remaining panicking tests fixed.

cyphar commented:
I've got all of the unit tests passing (there was a very minor screw up in my `selected_text` patch -- I set `is_word_selection = false` a bit too late in `onHoldPan` because the old code actually depended on when `selected_text` was set to non-nil, not when `selected_word` was set to nil). In any case, all of the unit tests now pass as well as some new unit tests I've added (I've also added some `assert.spy` checks to the language support hooks within the existing unit tests).

I would like to add some unit tests which specifically test Japanese text but because this requires a dictionary I wanted to ask before doing this -- there are free-as-in-freedom dictionaries available (JMdict being the main one) as well as public domain Japanese books (from Aozora Bunko). Would it be acceptable to add a dictionary and Japanese text to https://github.com/koreader/test-data (which would take up about 34MB)? I can then add some extra code to the testing framework to copy the Japanese dictionaries to `koreader/data/dict` during the Japanese EPUB tests.

I'm also porting over the Japanese deinflector unit tests from Yomichan (thought maybe it's not strictly necessary to port everything -- maybe just some smoke tests would be sufficient).

poire-z commented:
(I'm not friend with unit tests - so for me, no test is good enough :) but you added/updated quite a few of them for the "core" stuff, which is good enough. Any japanese plugin specific ones are up to you, afterall, it's _your_ territory now :)

Frenzie commented:
> Already done, unless you wanted something more specific than I already have?

Pardon, I misinterpreted your question then. Yes, that's basically what I was thinking. Since you mentioned something about multi-page selection at some point perhaps you could test that too, if you think it's worthwhile (and not too much of a bother/too complicated, not all tests are worth having ^_^).

> Ah yeah I opened it as a draft originally, sorry. I can re-open this as a new PR if it's necessary to get CI to trigger?

I doubt that's it, but let me try the old closing/reopening trick real quick.

poire-z commented:
> Ah yeah I opened it as a draft originally, sorry

Which was fine ! I'm not even sure that's the reason :)

> I can re-open this as a new PR if it's necessary to get CI to trigger?

May be not yet.
I just merged a PR, so may be you can try rebasing to the new master and force push - this might trigger it now that it is no longer marked as Draft ?

cyphar commented:
Ah oops, yeah will do. I'll try to not open the next PR as a draft, that might fix things?

Geniusssmit commented:
Thank you for this update! I have some problems with deconjugation. as you can see with some words dictionary returns random entries (please look at query).

![IMG_20211120_121136](https://user-images.githubusercontent.com/24941285/142730293-24139399-81dd-41d2-9e03-28b3431a138d.jpg)
![IMG_20211120_121203](https://user-images.githubusercontent.com/24941285/142730295-0c3272bb-fa14-4ffe-b2b6-f04e8316cba2.jpg)
![IMG_20211120_121446](https://user-images.githubusercontent.com/24941285/142730296-97f593d7-7b82-47dc-a593-b514e68f8052.jpg)
![IMG_20211120_121545](https://user-images.githubusercontent.com/24941285/142730297-a3cf1400-4933-4240-8ea1-6f018836a22c.jpg)
![IMG_20211120_121413](https://user-images.githubusercontent.com/24941285/142730298-d113801e-d042-4f46-adac-bafc920110da.jpg)
![IMG_20211120_121358](https://user-images.githubusercontent.com/24941285/142730299-f13383af-9d72-4e83-bbf3-34900f5d7c4a.jpg)

poire-z commented:
(Can't comment on the Japanese stuff, but just surprised with these screenshots showing `1/153` or `1/277`, meaning there's 277 results ! @cyphar : is that expected with your Japanese fuzziness ? :)

cyphar commented:
@Geniusssmit You need to disable fuzzy searching. Fuzzy searching does not work at all with Japanese text, but it's enabled by default because it's useful for other languages. (Maybe the Japanese plugin should disable fuzzy searching entirely but that seems a bit drastic and would probably causes issues.)

@poire-z Yes, though this was a problem before this plugin as well (I mention this issue in one of the issue threads -- searching 喋られる would give you ~100 results with fuzzy searching, none of which were useful).

poire-z commented:
> Fuzzy searching does not work at all with Japanese text, but it's enabled by default because it's useful for other languages

If you can detect the word is japanese (when your japanese support has done stuff) before invoking sdcv, couldn't you just one-shot-disable fuzzy search in that case ? So users reading both western and japanese can keep fuzzy search enabled.

Geniusssmit commented:
> @Geniusssmit You need to disable fuzzy searching (Lookup > Dictionary Settings). Fuzzy searching does not work at all with Japanese text, but it's enabled by default because it's useful for other languages. (Maybe the Japanese plugin should disable fuzzy searching entirely but that seems a bit drastic and would probably cause issues.) You might have noticed the issue is that you're getting results where only the kana match -- this is what stardict's fuzzy searching does with Japanese text (which is obviously broken for Japanese).
> 
> @poire-z Yes, though this was a problem before this plugin as well (I mention this issue in one of the issue threads -- searching 喋られる would give you ~100 results with fuzzy searching, none of which were useful).

Thanks, it worked. Btw I wonder how to change text orientation to vertical. Japanese literature is always written vertically. I'm glad that now looking up works but books should be vertical.

poire-z commented:
Vertical text is not supported, See #4353, https://github.com/koreader/koreader/issues/4353#issuecomment-452452412.

Just by curiosity, I think I asked somewhere but got no answer, so asking again:
With a long paragraph / string of japanese characters that would occupy more than one screen/page, when it's displayed vertically, how do web browsers and other EPUB renderers do ? How do they wrap the lines, what scrolling do they allow ?
A web browser doesn't make pages, so I guess it uses the viewport height to make a line, and will allow you to scroll horizontally to read the text that didn't fit on a page, right ? Or do authors  have to specify a height for each paragraph, and if that height is too tall, you'd have to scroll up and down to read a line? Is that practical to read ?
And EPUB renderers have to make pages... So, how do they cut pages ? When they have put some paragraph on a page, and the next one is too long for its text to fit on the remaining part of the page, are they clever enough to cut the linear japanese string to put the part that fit in the remaining vertical space of previous page, and the rest on a new page? Or do they always start on a new page ? Or do they expect paragraph to be small enough so they can be considered as a single full sized block (like an image), that will go fully on a page and won't be cut?
(Just asking cause I'm curious, but all this still feels heavily complicated and so different to be implemented in crengine.)

Geniusssmit commented:
@poire-z 
Sorry, I'm not sure how they do that. You can check it with https://ttu-ebook.web.app/ (good epub reader) it's not even cutting text into pages. also here is some file to test
[[高木レム] 絶対にメディア化できない小説.zip](https://github.com/koreader/koreader/files/7574915/default.zip)
Calibre's reader look's like that 
![image](https://user-images.githubusercontent.com/24941285/142735392-85f807c6-1c87-45ac-abfc-a4585c946a16.png)

cyphar commented:
@poire-z It'd probably be simpler to just disable fuzzy searching for CJK-containing words (detecting if a word is Japanese is far more complicated and has many more edge-cases). I suspect that fuzzy searching isn't particularly useful for Korean for the same reason it doesn't work with Japanese (on paper if sdcv understood Hangul it might work but I don't think it does).

> Just by curiosity, I think I asked somewhere but got no answer, so asking again: With a long paragraph / string of japanese characters that would occupy more than one screen/page, when it's displayed vertically, how do web browsers and other EPUB renderers do? How do they wrap the lines, what scrolling do they allow? A web browser doesn't make pages, so I guess it uses the viewport height to make a line, and will allow you to scroll horizontally to read the text that didn't fit on a page, right ? Or do authors have to specify a height for each paragraph, and if that height is too tall, you'd have to scroll up and down to read a line? Is that practical to read ?

The text is wrapped at the bottom of the screen in the same way that you wrap text horizontally (except vertical :wink:). You cannot scroll vertically with vertical text. And for the record -- vertical text is read top-down and right-to-left (meaning that page turning needs to be flipped -- just as with Arabic and Hebrew).

> And EPUB renderers have to make pages... So, how do they cut pages ? When they have put some paragraph on a page, and the next one is too long for its text to fit on the remaining part of the page, are they clever enough to cut the linear japanese string to put the part that fit in the remaining vertical space of previous page, and the rest on a new page? Or do they always start on a new page ? Or do they expect paragraph to be small enough so they can be considered as a single full sized block (like an image), that will go fully on a page and won't be cut?

The sentence or paragraph is cut the same way as you would with horizontal text. Outside of obeying the line breaking rules (which koreader already handles) you can split a sentence anywhere and continue it on the next line or the next page.

> (Just asking cause I'm curious, but all this still feels heavily complicated and so different to be implemented in crengine.)

As far as I know, the following readers are the only ones which handle vertical Japanese text as well as it is possible to do:

 * Calibre
 * Modern web browsers (so https://ttu-ebook.web.app/).
 * Kindle's and Kobo's readers.
 * Readily (Android app)

I suspect that if you try playing around with a vertical Japanese book with one of the above readers, you'll be able to get far better answers than I can give (since I've never worked with crengine's text layout code, so I'm not sure what questions you might want answered).

I was hoping to look at vertical text in the future when I have a little more time (I spent a fair amount of my "studying Japanese" time working on these various PRs so I'll probably need to take a bit of a break from koreader contributions).

poire-z commented:
> so I'm not sure what questions you might want answered

Tried  https://ttu-ebook.web.app/ with the provided EPUB, and OK, I get that in a fully TBRL (Top-Bottm Right-Left) book, you should get a big horizontal "roll" and you cut page vertically when you can (instead of the big vertical roll that we get with western text). So, in this context, you would just cut vertical slices in a bit horizontal roll.
I just had a hard time making this shift in my mind :)
So, my question (my mind staying my western mind) is rather about what would / should happen when TBRL/TBLR paragraphs happen in a normal western LRTB book, with pages cut by making horizontal slices in a bit vertical roll : 

![image](https://user-images.githubusercontent.com/24273478/142742366-49b8d7a4-add2-4e2d-9ecf-5adbf5bcb765.png)

I wonder how renderers would do with this yellow TBRL paragraph.
My brain is also hardcoded on how crengine works, rendering all block individually, stacking them vertically without any notion of pages - and then a last pass is made to figure out where to cut page in this vertical stacking. With the above picture, you'd need info about the pages (what height is available on the 2nd page to figure out where to cut a line in the yellow block) when doing the first step - which is just impossible in the way crengine is made.

cyphar commented:
> So, my question (my mind staying my western mind) is rather about what would / should happen when TBRL/TBLR paragraphs happen in a normal western LRTB book, with pages cut by making horizontal slices in a bit vertical roll :

That sounds like a nightmare :smile:. I suspect the only sane thing to do (if we wanted to make it seamless) is to treat whatever remaining vertical space (after the LRTB text) as if it were the entire page and then lay out the TBRL subsection normally (wrapping at the height and then splitting over pages where necessary).

However I suspect if this were to be practically implemented in KOReader it would be better to just split the differently-oriented text onto separate pages because such text would be exceptionally uncommon and thus not worth over-complicating crengine for (in my experience, Japanese text mixed with English text always uses the same writing direction for both -- whether they use LRTB or TBRL depends on the medium and how much English text there is).

cyphar commented:
#8458 disables fuzzy searching for CJK word lookups.

rtega commented:
Do I read it correctly that this doesn't work yet on android? I tried the last release and it seems to select minimal sections (say one or two characters) instead of bigger sections (several characters) as yomichan does.

Frenzie commented:
> Do I read it correctly that this doesn't work yet on android?

Not sure where you got that idea, but no. ;-)

cyphar commented:
It definitely works on Android, I've been using it for the past few weeks. If you believe you've found an issue please open a separate issue.

poire-z commented:
I see this logged a lot on my side:
https://github.com/koreader/koreader/blob/eda8379e8cd07418f51c92b6a0d1f962616acb21/frontend/apps/reader/modules/readerdictionary.lua#L795-L797

I think it's for the edge case (that is possibly mine only) of me having a secondary dict_ext dict dir:
https://github.com/koreader/koreader/blob/eda8379e8cd07418f51c92b6a0d1f962616acb21/frontend/apps/reader/modules/readerdictionary.lua#L730-L732
with:
https://github.com/koreader/koreader/blob/eda8379e8cd07418f51c92b6a0d1f962616acb21/frontend/apps/reader/modules/readerdictionary.lua#L714-L725

So, I may get 2 results with 2 queries for a single word. Always 2 I guess.
How do you feel this should be fixed ? 
`if #all_results ~= #words * #dict_dirs then` ?
`if #all_results <= #words * #dict_dirs then` ?
Or just removing this check and warning ?

cyphar commented:
The warning is important because having results like that could confuse users of rawSdcv (namely the Japanese plugin which relies on the output array from sdcv matching the words passed to rawScdv).

If the check is wrong for dict_ext (which it seems like it is), I will need to take a look to figure out how to handle this so that the Japanese support plugin doesn't break with dict_ext dictionaries...

poire-z commented:
> Maybe we should merge dictionary results in rawSdcv if they correspond to the same lookup word (in other words, make dict_ext transparent to the caller).

That's explicitely not what I want :) I want dict_ext results to be the last ones (or sole ones, when dict/ don't have any results), as if dict/ has results, they are better than dict_ext/ ones.

May be for your japanese plugins, you should explicitely request to check only dict/ and forget about dict_ext ?

cyphar commented:
> May be for your japanese plugins, you should explicitely request to check only dict/ and forget about dict_ext ?

That would also work, though it would cause some issues if someone's preferred Japanese dictionary is in dict_ext but I guess that's a pretty small edge case (in that case we could handle it by calling rawSdcv twice but probably better to not over-complicate things unless it's truly necessary).

poire-z commented:
I'm probably the only person using this dict_ext feature - so yes, don't complicate things too much.
(May be it can have some use case with Japanese that I'm too foreign to imagine :)

Frenzie commented:
I don't use the dictionary feature on my device much period. I do most of my reading at home and I prefer the serendipity afforded by my physical Petit Robert or Petit Larousse. When I do go on a longer train ride or something I tend to prefer to use my phone on the side. Doing it all in one doesn't have my preference, but I appreciate it's there.[1] The dict_ext thing is something that's very useful in theory but indeed probably not used much in practice. :-)

[1] The Wikipedia display you made a few years ago is excellent btw.

poire-z commented:
> > Maybe we should merge dictionary results in rawSdcv if they correspond to the same lookup word (in other words, make dict_ext transparent to the caller).
> 
> That's explicitely not what I want :) I want dict_ext results to be the last ones (or sole ones, when dict/ don't have any results), as if dict/ has results, they are better than dict_ext/ ones.

I think I did not get what you meant with "merge dictionary results" at the time, but now I think I do :)
Was that what you had in mind ?

```diff
--- a/frontend/apps/reader/modules/readerdictionary.lua
+++ b/frontend/apps/reader/modules/readerdictionary.lua
@@ -776,25 +776,30 @@ function ReaderDictionary:rawSdcv(words, dict_names, fuzzy_search, lookup_progre
             -- so it's safe to split. Ideally luajson would support jsonl but
             -- unfortunately it doesn't and it also seems to decode the last
             -- object rather than the first one if there are multiple.
+            local sdcv_run_num = 0
             for _, entry_str in ipairs(util.splitToArray(results_str, "\n")) do
+                sdcv_run_num = sdcv_run_num + 1
                 local ok, results = pcall(JSON.decode, entry_str)
-                if ok and results then
-                    table.insert(all_results, results)
-                else
+                if not ok or not results then
                     logger.warn("JSON data cannot be decoded", results)
                     -- Need to insert an empty table so that the word entries
                     -- match up to the result entries (so that callers can
                     -- batch lookups to reduce the cost of bulk lookups while
                     -- still being able to figure out which lookup came from
                     -- which word).
-                    table.insert(all_results, {})
+                    results = {}
                 end
+                if all_results[sdcv_run_num] then
+                    util.arrayAppend(all_results[sdcv_run_num], results)
+                else
+                    table.insert(all_results, results)
+                end
+            end
+            if #all_results ~= #words then
+                logger.warn("sdcv returned a different number of results than the number of words")
             end
         end
     end
-    if #all_results ~= #words then
-        logger.warn("sdcv returned a different number of results than the number of words")
-    end
     return lookup_cancelled, all_results
 end

```
It works for my use case (no warning, results from both dicts combined and in the expected order).
I think it should in your japanese plugin use case too, right ? Can you give it a quick test (at least, when only with a single dict/ dir) that it does not break anything.
I'll make a PR if it works for you (and if you didn't envision that something more complex is needed).

cyphar commented:
@poire-z That is sort of what I meant -- though `sdcv_run_num` is a bit of a confusing name because it has nothing to do with which `sdcv` execution is being handled (I had to re-read the patch a few times to figure that out) -- it should be `sdcv_result_num` or just `result_word_idx` or something to indicate that it's the index of the word passed to sdcv and index of the list of results returned by sdcv.

poire-z commented:
> though sdcv_run_num is a bit of a confusing name

Indeed :/ It kept the name from some earlier iteration of my attempt at making it right :) Will rename.

> Also the if #all_results ~= #words can be kept where it is -- you don't need to check it with each loop.

Yes. But I thought we might want to check this for EACH sdcv run (otherwise, one good run might mask the other run failure to give the good nb of results). So, it we keep it inside the loop, it should use `sdcv_run_num` (or its new name) instead of `#all_results`.

So, does this read better?:

```diff
--- a/frontend/apps/reader/modules/readerdictionary.lua
+++ b/frontend/apps/reader/modules/readerdictionary.lua
@@ -776,25 +776,30 @@ function ReaderDictionary:rawSdcv(words, dict_names, fuzzy_search, lookup_progre
             -- so it's safe to split. Ideally luajson would support jsonl but
             -- unfortunately it doesn't and it also seems to decode the last
             -- object rather than the first one if there are multiple.
+            local result_word_idx = 0
             for _, entry_str in ipairs(util.splitToArray(results_str, "\n")) do
+                result_word_idx = result_word_idx + 1
                 local ok, results = pcall(JSON.decode, entry_str)
-                if ok and results then
-                    table.insert(all_results, results)
-                else
+                if not ok or not results then
                     logger.warn("JSON data cannot be decoded", results)
                     -- Need to insert an empty table so that the word entries
                     -- match up to the result entries (so that callers can
                     -- batch lookups to reduce the cost of bulk lookups while
                     -- still being able to figure out which lookup came from
                     -- which word).
-                    table.insert(all_results, {})
+                    results = {}
                 end
+                if all_results[result_word_idx] then
+                    util.arrayAppend(all_results[result_word_idx], results)
+                else
+                    table.insert(all_results, results)
+                end
+            end
+            if result_word_idx ~= #words then
+                logger.warn("sdcv returned a different number of results than the number of words")
             end
         end
     end
-    if #all_results ~= #words then
-        logger.warn("sdcv returned a different number of results than the number of words")
-    end
     return lookup_cancelled, all_results
 end
```

cyphar commented:
@poire-z Good point, and yeah that looks reasonable. Unfortunately I haven't been able to get the emulator to build on my machine for the past few weeks (a super cryptic CMake error about ccache not being a working C compiler which I couldn't fix by doing clean rebuilds or even a fresh clone of the project) so I can't test this at the moment.

j9d3it commented:
Hello, I'm trying to use this plugin but am having trouble. I downloaded some Japanese news using the News downloader. If I look at the book information, the Language is "N/A", rather than Japanese. Does the plugin have to use this value to be enabled? Is there a way to force-enable the plugin?

cyphar commented:
@j9d3it No the document language doesn't need to be Japanese. You just need to have a Japanese dictionary of some kind installed.

Frenzie commented on plugins/japanese.koplugin/deinflector.lua:
```diff
@@ -0,0 +1,470 @@
+                    nvariants = nvariants + #rules
+                    nrules = nrules + 1
+                end
+                local nrules_str = T(N_("%1 rule", "%1 rules", nrules), nrules)
+                local nvariants_str = T(N_("%1 variant", "%1 variants", nvariants), nvariants)
+                UIManager:show(InfoMessage:new{
+                    -- @translators %1 is the "%1 rule(s)" string, %2 is the "%1 variant(s)" string.
+                    text = T(_("Deinflector has %1 and %2 loaded."), nrules_str, nvariants_str),
```
@cyphar There are some complaints about this construction.

![image](https://user-images.githubusercontent.com/202757/222892825-674a83a8-ca26-479c-8e18-0c163f5677ea.png)

<https://hosted.weblate.org/translate/koreader/koreader/en/?checksum=66e64e8b6763628e>

cyphar replied:
The issue is that there are two numerical values in the string which need to be pluralised, and (at least when I checked while writing this) you cannot have more than one variables in a single string with pluralisation. While the values are never going to be 0 or 1, some languages have complicated pluralising rules so just combining them into one string would also make translations difficult or impossible.

The only way I can see to make this easier to translate is to make it two sentences instead (though it would make the info screen more wordy)?

Frenzie replied:
Yes, without having two separate sentences it can be done in GetText, only in Fluent or some such.

Or alternatively something like

Deinflector loaded:
 * 1 rule(s)
 * 1 variant(s)
